data_safe_haven/external/api/graph_api.py

```diff
@@ -225,43 +230,32 @@ def create_application(
                 self.logger.info(
                     f"Created new application '[green]{json_response['displayName']}[/]'.",
                 )
+
+            # Ensure that the application service principal exists
+            self.ensure_application_service_principal(application_name)
+
             # Grant admin consent for the requested scopes
             if application_scopes or delegated_scopes:
-                application_id = self.get_id_from_application_name(application_name)
-                application_sp = self.get_service_principal_by_name(application_name)
-                if not (
-                    application_sp
-                    and self.read_application_permissions(application_sp["id"])
-                ):
-                    self.logger.info(
-                        f"Application [green]{application_name}[/] has requested permissions"
-                        " that need administrator approval."
-                    )
-                    self.logger.info(
-                        "Please sign-in with [bold]global administrator[/] credentials for the"
-                        " Azure Active Directory where your users are stored."
-                    )
-                    self.logger.info(
-                        "To sign in, use a web browser to open the page"
-                        f" [green]https://login.microsoftonline.com/{self.tenant_id}/adminconsent?client_id="
-                        f"{application_id}&redirect_uri=https://login.microsoftonline.com/common/oauth2/nativeclient[/]"
-                        " and follow the instructions."
-                    )
-                    while True:
-                        if application_sp := self.get_service_principal_by_name(
-                            application_name
-                        ):
-                            if self.read_application_permissions(application_sp["id"]):
-                                break
-                        time.sleep(10)
+                for scope in application_scopes:
+                    self.grant_application_role_permissions(application_name, scope)
+                for scope in delegated_scopes:
+                    self.grant_delegated_role_permissions(application_name, scope)
+                while True:
+                    if application_sp := self.get_service_principal_by_name(
+                        application_name
+                    ):
+                        if self.read_application_permissions(application_sp["id"]):
+                            break
+                    time.sleep(10)
```

Comment:
```suggestion
                attempts = 0
                max_attempts = 5
                while attempts < max_attempts:
                    if application_sp := self.get_service_principal_by_name(
                        application_name
                    ):
                        if self.read_application_permissions(application_sp["id"]):
                            break
                    time.sleep(10)
                    attempts += 1
                
                if attempts == max_attempts:
                    raise DataSafeHavenException("Maximum attempts to validate service principle permissions exceeded")
```

Infinite loops feel like a bad idea.
Could we do this an raise an appropriate exception?

Reply:
Is this resolved?